nameresolution/sqlite/sqlite.go

```diff
@@ -0,0 +1,317 @@
+		s.logger.Warn("Configuring name resolution with an in-memory SQLite database. Service invocation across different apps will not work.")
+	} else {
+		s.logger.Infof("Configuring SQLite name resolution with path %s", connString[len("file:"):strings.Index(connString, "?")])
+	}
+
+	s.db, err = sql.Open("sqlite", connString)
+	if err != nil {
+		return fmt.Errorf("failed to create connection: %w", err)
+	}
+
+	// Performs migrations
+	err = performMigrations(ctx, s.db, s.logger, migrationOptions{
+		HostsTableName:    s.metadata.TableName,
+		MetadataTableName: s.metadata.MetadataTableName,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to perform migrations: %w", err)
+	}
+
+	// Init the background GC
+	err = s.initGC()
+	if err != nil {
+		return err
+	}
+
+	// Register the host and update in background
+	err = s.registerHost(ctx)
+	if err != nil {
+		return err
+	}
+
+	s.wg.Add(1)
+	go s.renewRegistration()
+
+	return nil
+}
+
+func (s *resolver) initGC() (err error) {
+	s.gc, err = internalsql.ScheduleGarbageCollector(internalsql.GCOptions{
+		Logger: s.logger,
+		UpdateLastCleanupQuery: func(arg any) (string, any) {
+			return fmt.Sprintf(`INSERT INTO %s (key, value)
+				VALUES ('nr-last-cleanup', CURRENT_TIMESTAMP)
+				ON CONFLICT (key)
+				DO UPDATE SET value = CURRENT_TIMESTAMP
+					WHERE unixepoch(CURRENT_TIMESTAMP) - unixepoch(value)  > ?;`,
+				s.metadata.MetadataTableName,
+			), arg
+		},
+		DeleteExpiredValuesQuery: fmt.Sprintf(
+			`DELETE FROM %s WHERE unixepoch(CURRENT_TIMESTAMP) - last_update < %d`,
+			s.metadata.TableName,
+			int(s.metadata.UpdateInterval.Seconds()),
+		),
+		CleanupInterval: s.metadata.CleanupInterval,
+		DB:              internalsql.AdaptDatabaseSQLConn(s.db),
+	})
+	return err
+}
+
+// Registers the host
+func (s *resolver) registerHost(ctx context.Context) error {
+	// Get the registration ID
+	u, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("failed to generate registration ID: %w", err)
+	}
+	s.registrationID = u.String()
+
+	queryCtx, queryCancel := context.WithTimeout(ctx, s.metadata.Timeout)
+	defer queryCancel()
+
+	// There's a unique index on address
+	// We use REPLACE to take over any previous registration for that address
+	// TODO: Add support for namespacing. See https://github.com/dapr/components-contrib/issues/3179
+	_, err = s.db.ExecContext(queryCtx,
+		fmt.Sprintf("REPLACE INTO %s (registration_id, address, app_id, namespace, last_update) VALUES (?, ?, ?, ?, unixepoch(CURRENT_TIMESTAMP))", s.metadata.TableName),
+		s.registrationID, s.metadata.GetAddress(), s.metadata.appID, "",
+	)
+	if err != nil {
+		return fmt.Errorf("failed to register host: %w", err)
+	}
+
+	return nil
+}
+
+// In backgrounds, periodically renews the host's registration
+// Should be invoked in a background goroutine
+func (s *resolver) renewRegistration() {
+	defer s.wg.Done()
+
+	addr := s.metadata.GetAddress()
+
+	d := s.metadata.UpdateInterval - s.metadata.Timeout
```

Comment:
We are we subtracting the Timeout here?

Reply:
The status must be updated every `UpdateInterval`, let's say 10s. If the table hasn't been updated in 10s, callers querying it will consider the record expired. However, queries are given a `Timeout`, for example 1s. So if we were to just loop every `UpdateInterval`, and the query did take 1s, it would be updated in 11s.

There's a check in the metadata struct that ensures that `UpdateInterval > Timeout`.

Reply:
Should we not have instances consider a record to be stale after a longer period than the `UpdateInterval`? Having the update interval the same as the TTL of the record means that there is no slack in the time windows and inherently leads to hosts being thrashed from the registry.

Reply:
I'm not sure I understand, sorry.

Currently we consider as stale after `UpdateInterval`, but we update the record every `UpdateInterval - Timeout`. (I am thinking of adding a `-1` there too as well). So there is some natural slack there?

Reply:
If we consider `Timeout` to be a very small number (basically 0), then there is a race condition whereby one instance reads and one writes _just after_ `UpdateInterval`. If the read was first then the reader would consider the host record to be stale. We can alleviate this issue by considering a record stale when it is `1.3*` past the `UpdateInterval`.

Reply:
Adding a `-1` to the timeout should be enough to solve this (and making sure that `UpdateInterval > Timeout + 1`). There shouldn't be race conditions in that case, since the timeout controls the query timeout too.

Reply:
Updated